providers/base/bin/zapper_keyboard_test.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
+"""
+This program tests whether the system responds correctly to an external
+USB keyboard.
+
+To run the test you need Zapper board connected and set up.
+"""
+import os
+import struct
+import sys
+import threading
+from enum import Enum
+
+from checkbox_support.scripts.zapper_proxy import zapper_run  # noqa: E402
+
+ZAPPER_KBD = "usb-Canonical_Zapper_main_board_123456-event-kbd"
+EVENT_BIN_FORMAT = "llHHI"
+
+ROBOT_TESTCASE_COMBO = """
+*** Settings ***
+Library    libraries/ZapperHid.py
+
+*** Test Cases ***
+Press and Release a combination of keys.
+    ${keys}=    Create List     A   B
+    Keys Combo    ${keys}
+"""
+
+ROBOT_TESTCASE_TYPE = """
+*** Settings ***
+Library    libraries/ZapperHid.py
+
+*** Test Cases ***
+Type the requested string
+    Type String    hello
+"""
+
+
+class KeyEvent(Enum):
+    UP = 0
+    DOWN = 1
+
+
+def _listen_keyboard_events(event_file_path, callback):
+    """Listen for keyboard events on the given file."""
+    with open(event_file_path, "rb") as event:
+        while True:
+            read_keyboard_events(event, callback)
+
+
+def read_keyboard_events(event, callback):
+    """Read keyboard events from the given file and run the callback."""
+    data = event.read(struct.calcsize(EVENT_BIN_FORMAT))
+    _, _, event_type, code, value = struct.unpack(EVENT_BIN_FORMAT, data)
+    if event_type == 1:  # 0x01 is for _kbd_ events
+        callback((KeyEvent(value), code))
+
+
+def assert_key_combo(host, events):
+    """Request to press a key combination and expect the corresponding events."""
+
+    events.clear()
+    zapper_run(host, "robot_run", ROBOT_TESTCASE_COMBO.encode(), {}, {})
+
+    assert events == [
+        (KeyEvent.DOWN, 30),
+        (KeyEvent.DOWN, 48),
+        (KeyEvent.UP, 48),
+        (KeyEvent.UP, 30),
+    ]
+
+
+def assert_type_string(host, events):
+    """Request to type a string and expect the corresponding events."""
+
+    events.clear()
+    zapper_run(host, "robot_run", ROBOT_TESTCASE_TYPE.encode(), {}, {})
+
+    assert events == [
+        (KeyEvent.DOWN, 35),
+        (KeyEvent.UP, 35),
+        (KeyEvent.DOWN, 18),
+        (KeyEvent.UP, 18),
+        (KeyEvent.DOWN, 38),
+        (KeyEvent.UP, 38),
+        (KeyEvent.DOWN, 38),
+        (KeyEvent.UP, 38),
+        (KeyEvent.DOWN, 24),
+        (KeyEvent.UP, 24),
+    ]
+
+
+def main(argv):
+    if len(argv) != 2:
+        raise SystemExit("Usage: {} <zapper-ip>".format(argv[0]))
+
+    zapper_kbd_path = f"/dev/input/by-id/{ZAPPER_KBD}"
```

Comment:
This dodges a problem where the FS node doesn't exist, but will still yield exceptions when there's a problem when opening the file (permissions or other IO errors).
So following a more pythonic EAFP instead of LBYL can give you better results here.
Namely surrounding the actual opening code with a try/except block would potentially capture more classes of errors and also you could use `raise SystemExit(...) from exc` that facilitates debugging.

Reply:
I want to open this file in the listener instead, it makes the main() function simpler. I'm now using `os.access` to check for file existence and permission.